<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Kan Etoh" />


<title>RNAseqChef manual</title>

<script src="site_libs/header-attrs-2.16/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore">RNAseqChef manual</h1>
<h4 class="author">Kan Etoh</h4>
<h4 class="date">22/09/30 13:54</h4>

</div>


<hr />
<p><a href="https://doi.org/10.5281/zenodo.7095218"><img
src="https://zenodo.org/badge/DOI/10.5281/zenodo.7095218.svg"
alt="DOI" /></a><br></p>
<div id="outline" class="section level1">
<h1>Outline</h1>
<p><img src="manual_figure/graphical%20abstract_ver2.png" id="id"
class="class" style="width:70.0%" /> <br><br></p>
<p>RNAseqChef, RNA-seq data controller highlighting gene expression
feature, a web-based platform of systematic transcriptome analysis which
can automatically detect, integrate, and visualize the differentially
expressed genes and their biological function without bioinformatics
skills. Furthermore, Users can analyze not only a single dataset but
also integrate and evaluate multiple datasets.<br> You can download a
publication-ready high resolution figure (PDF file that can be edited by
Illustrator).<br><br> RNAseqChef can be used in the following way:<br>
1. Access <a href="https://imeg-ku.shinyapps.io/RNAseqChef/">the web
page</a> <br> 2. Execute the code described in “Local installation” on
<a href="https://github.com/Kan-E/RNAseqChef">the source code page</a>
in R <br><br> Web page: <a
href="https://imeg-ku.shinyapps.io/RNAseqChef/"
class="uri">https://imeg-ku.shinyapps.io/RNAseqChef/</a> <br> Source
code: <a href="https://github.com/Kan-E/RNAseqChef"
class="uri">https://github.com/Kan-E/RNAseqChef</a> <br> Japanese
manual: <a href="https://kan-e.github.io/RNAseqChef_manual_japanese"
class="uri">https://kan-e.github.io/RNAseqChef_manual_japanese</a> <br>
Please contact me: Kan Etoh, <a href="mailto:kaneto@kumamoto-u.ac.jp"
class="email">kaneto@kumamoto-u.ac.jp</a> <br><br><br></p>
<p>RNAseqChef consists of the following sections:<br></p>
<ul>
<li><p><strong>Pair-wise DEG</strong> (Differential expression analysis
for pair-wise comparison)</p></li>
<li><p><strong>3 conditions DEG</strong> (Differential expression
analysis for 3 conditions multiple comparison)</p></li>
<li><p><strong>Multi DEG</strong> (Differential expression analysis for
multiple comparison by likelihood ratio test)</p></li>
<li><p><strong>Venn diagram</strong> (Visualization of the intersections
of gene lists)</p></li>
<li><p><strong>Normalized count analysis</strong> (Clustering analysis
for normalized count data)</p></li>
<li><p><strong>Enrichment viewer</strong> (Functional analysis of gene
cluster list)</p></li>
<li><p><strong>More</strong>（Other useful functions, Reference,
etc）</p></li>
</ul>
<p>To establish the utility, flexibility and versatility of RNAseqChef,
we designed to divide the core functions into 6 sections, which enables
us to handle various types of input files. <br> The user-interface
consists of <strong>Menu</strong>, <strong>Setting panel</strong>,
<strong>Tab panel</strong>.<br> Each section can be moved by clicking
the menu button.<br> Input files can be uploaded in the setting
panel.<br> RNA-seq analysis is automatically started by clicking the tab
panel in the output panel.<br> Please explore the app’s features with
the example data by clicking <strong>the example button</strong> in the
setting panel.<br><br> <img src="manual_figure/panel.png" class="class"
style="width:70.0%" alt="panel" /> <br></p>
<hr />
</div>
<div id="common-rules-and-important-notice-for-input-file"
class="section level1">
<h1>Common rules and important notice for input file</h1>
<div id="file-extension" class="section level3">
<h3>File extension</h3>
<p>Tab-separated txt file (.txt), comma-separated csv file (.csv), and
excel file (.xlsx) are accepted.<br><br></p>
</div>
<div id="duplication-of-gene-names" class="section level3">
<h3>Duplication of gene names</h3>
<p>Be careful with the autocorrect function of Excel that causes gene
name errors.<br> Depending on the version of Excel, <strong>gene name
conversion/duplication occurs just by opening the file</strong>.（For
example, March1 → 1-Mar, Marc1 → 1-Mar）<br> RNAseqChef cannot handle
files containing duplicate gene names.<br> If you edit the file in
Excel, ensure no misconversions in the gene names.<br><br></p>
<hr />
</div>
</div>
<div id="common-rules-and-important-notice-for-output-file"
class="section level1">
<h1>Common rules and important notice for output file</h1>
<div id="data-save" class="section level3">
<h3>Data save</h3>
<p>By pressing the download button, the table data and graph will be
saved. Table data is saved as a tab-separated txt file (.txt), and
graphs are saved as a PDF file that can be edited by Adobe
Illustrator.<br> Only in batch mode of pair-wise DEG, multiple files are
saved in zip format.<br><br></p>
</div>
<div id="adjusting-the-plot-size" class="section level3">
<h3>Adjusting the plot size</h3>
<p>You can change the figure size by adjusting “Output plot size setting
for pdf” in the Setting panel.<br> You can adjust the parameters when
adjacent characters overlap in the graph or when the margins of the
graph are too large.<br> If the value of the parameter is 0, it will be
saved in the default size.<br> The default size depends on the type of
graph. For details, please see the pop-up in the app.<br><br></p>
<hr />
</div>
</div>
<div id="pair-wise-deg" class="section level1">
<h1>Pair-wise DEG</h1>
<p><strong>Pair-wise DEG</strong> detects and visualizes differentially
expressed genes. This function executes a series of pair-wise DEG
analyses such as “DEG (differentially expressed gene) detection”,
“clustering analysis”, and “enrichment analysis”.</p>
<div id="setting" class="section level2">
<h2>Setting</h2>
<p><img src="manual_figure/pair-wise%20set.png" class="class"
style="width:50.0%" alt="pair-wise set" /><br></p>
<hr />
<div id="input-format" class="section level3">
<h3>Input format</h3>
<p>Three types of raw count data formats can be used as input.<br> By
clicking the button as shown below, the upload format can be
switched.<br> <img src="manual_figure/input%20switch.png" class="class"
style="width:100.0%" alt="input switch" /><br></p>
<div id="raw-count-data" class="section level4">
<h4>1. <strong>raw count data</strong> <br></h4>
<p>Analysis can only be performed with raw count data if the following
conditions are fulfilled:</p>
<pre><code>- A file consists of only two groups of data to be analyzed.
- The replication number is represented by the underline “_”.
- Do not use the underline &quot;_&quot; for anything else. </code></pre>
<p><img src="www/input_format1.png" class="class" style="width:50.0%"
alt="unchanged image" /><br>　</p>
<hr />
</div>
<div id="raw-count-data-metadata" class="section level4">
<h4>2. <strong>raw count data + metadata</strong> <br></h4>
<p>You can use this format if the above conditions are not fulfilled,
for example, if the sample name is an accession number or if the raw
count data contains extra information that is not the subject of
analysis.<br> Metadata must contain the following information:</p>
<pre><code>- The first column is the sample name used in the raw count data.（ex. accession number）
- The second column is the corresponding sample name that matches the sample name in the first column. (e.g. Control_1)
- The third and subsequent columns do not affect the analysis.</code></pre>
<p><img src="www/input_format2.png" class="class" style="width:50.0%"
alt="unchanged image" /> <br><br></p>
</div>
<div id="raw-count-data-batch-mode" class="section level4">
<h4>3. <strong>raw count data (Batch mode)</strong> <br></h4>
<p>Batch processing can only be performed with raw count data if the
following conditions are fulfilled:</p>
<ul>
<li>A file consists of only two groups of data to be analyzed. - The
replication number is represented by the underline “<em>”. - Do not use
the underline ”</em>” for anything else. In this batch mode, result data
can not be displayed in the output panel.<br />
You can get the zip files including Clustering, MA_plot, DEG_result, DEG
count data (up and down), and normalized count data, by clicking the
download button. <img src="www/input_format1.png" class="class"
style="width:50.0%" alt="unchanged image" /> <br>　</li>
</ul>
<hr />
</div>
<div id="deg-analysis-method" class="section level4">
<h4>DEG analysis method</h4>
<p>The following three types of DEG analysis can be selected.</p>
<ul>
<li><a
href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">DESeq2</a></li>
<li><a
href="https://academic.oup.com/bioinformatics/article/26/1/139/182458">edgeR</a></li>
<li><a
href="https://academic.oup.com/bioinformatics/article/29/8/1035/228913">EBSeq</a>
<br></li>
</ul>
</div>
<div id="fdr-false-discovery-rate-method" class="section level4">
<h4>FDR, False Discovery Rate, method</h4>
<p>The following three types of FDR methods can be selected if the DEG
analysis method is DESeq2 or edgeR.</p>
<ul>
<li>BH, Benjamini-Hochberg method (default)</li>
<li>Qvalue, Storey method</li>
<li><a href="https://www.nature.com/articles/nmeth.3885">IHW,
Independent Hypothesis Weighting</a> <br></li>
</ul>
</div>
<div id="species" class="section level4">
<h4>Species</h4>
<p>The following analysis is performed by selecting the dataset
species.</p>
<pre><code>- Conversion to gene symbol if the gene name is ENSEMBL ID
- Performing enrichment analysis</code></pre>
<p><br></p>
</div>
<div id="cut-off-conditions" class="section level4">
<h4>Cut-off conditions</h4>
<p>Three types of thresholds, <strong>Fold change</strong>,
<strong>FDR</strong>, and <strong>base mean</strong>, can be set. <br>
<br></p>
</div>
<div id="option-normalized-count-input" class="section level4">
<h4>Option: normalized count input</h4>
<p>Base mean cut-off can be set with the uploaded normalized count data
such as TPM counts.<br> The y-axis of the boxplot can be displayed with
the uploaded normalized count data.<br> Note: Uploading raw count data
is not unnecessary. <br> <br></p>
<hr />
</div>
</div>
</div>
<div id="output" class="section level2">
<h2>Output</h2>
<div id="input-data" class="section level3">
<h3>Input Data</h3>
<p>The uploaded raw count data is displayed.<br> In the case of the “Raw
count data + metadata” format, it also displays the raw count data that
is redefined using the uploaded metadata. <br><br> <img
src="manual_figure/pair-wise%20input.png" class="class"
style="width:70.0%" alt="pair-wise input" /><br><br></p>
<hr />
</div>
<div id="result-overview" class="section level3">
<h3>Result overview</h3>
<p>Three types of clustering analysis are performed: Principal Component
Analysis (PCA), Multi-Dimensional Scaling (MDS), and hierarchical
clustering with the ward.D2.<br> The DEG analysis that is selected in
the setting panel is performed to detect DEGs. MA-plot and heatmap are
displayed as a result of the DEG analysis.<br> The table data of the DEG
analysis is displayed.<br> <img
src="manual_figure/pair_result_overview.png" class="class"
style="width:80.0%" alt="pair-wise result_overview" /><br></p>
<hr />
</div>
<div id="goi-genes-of-interest-profiling" class="section level3">
<h3>GOI (Genes of interest) profiling</h3>
<p>The volcano plot is displayed.<br> By selecting genes from the GOI
list, you can label only the GOI among the points displayed in the
volcano plot.<br> The x-axis and y-axis ranges can be freely changed by
operating the slide bar.<br> The heatmap and boxplot of the genes
selected from the GOI list are displayed.<br><br> <img
src="manual_figure/pair-wise%20GOI%20profiling1.png" class="class"
style="width:80.0%" alt="pair-wise GOI profiling" /><br> <img
src="manual_figure/pair-wise%20GOI%20profiling2.png" class="class"
style="width:80.0%" alt="pair-wise GOI profiling" /> <br></p>
<hr />
</div>
<div id="enrichment-analysis" class="section level3">
<h3>Enrichment analysis</h3>
<p>Over-representation analysis and GSEA, Gene set enrichment analysis,
are performed to identify the function of DEGs.<br> The gene set can be
selected from the following: <br><a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=H">MSigDB
hallmark gene</a> <br> <a href="https://www.kegg.jp">KEGG</a> <br> <a
href="https://reactome.org">Reactome</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/resource/Pathway+Interaction+Database">PID
(Pathway Interaction Database)</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/dataset/Biocarta+Pathways">BioCarta</a>
<br> <a
href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a>
<br> <a href="http://geneontology.org">GO (biological process, cellular
component, and molecular function)</a> <br> <a
href="https://hpo.jax.org/app/">Human phenotype ontology</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(activator)</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(repressor)</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=TFT">Transcription
factor targets</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=MIR">miRNA
target</a> <br></p>
<p>Dotplot, GSEA plot, and gene-concept network, cnet, plot are
displayed as the result. <br> Enrichment analysis depends on <a
href="https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html">ClusterProfiler</a>.
<br> <img src="manual_figure/pair-wise%20enrichment%20analysis.png"
class="class" style="width:80.0%"
alt="pair-wise enrichment analysis" /><br></p>
<hr />
</div>
</div>
</div>
<div id="conditions-deg" class="section level1">
<h1>3 conditions DEG</h1>
<p>3 conditions DEG detects and visualizes differentially expressed
genes by EBSeq multi-comparison analysis.<br> Note: Multiple comparison
analysis takes 5-10 minutes.</p>
<div id="setting-1" class="section level2">
<h2>Setting</h2>
<p><img src="manual_figure/cond%203%20set.png" class="class"
style="width:50.0%" alt="cond 3 set" /><br></p>
<hr />
<div id="input-format-1" class="section level3">
<h3>Input format</h3>
<p>The input format and settings for 3conditions DEG are the same as for
pair-wise DEG.<br> If you have read the description of pair-wise DEG,
you do not need to read the below.<br><br></p>
<p>Two types of raw count data formats can be used as input.<br></p>
<div id="raw-count-data-1" class="section level4">
<h4>1. <strong>raw count data</strong> <br></h4>
<p>Analysis can only be performed with raw count data if the following
conditions are fulfilled:</p>
<pre><code>- A file consists of only three groups of data to be analyzed.
- The replication number is represented by the underline “_”.
- Do not use the underline &quot;_&quot; for anything else. </code></pre>
<p><img src="www/input_format1.png" class="class" style="width:50.0%"
alt="unchanged image" /> <br>　</p>
<hr />
</div>
<div id="raw-count-data-metadata-1" class="section level4">
<h4>2. <strong>raw count data + metadata</strong> <br></h4>
<p>You can use this format if the above conditions are not fulfilled,
for example, if the sample name is an accession number or if the raw
count data contains extra information that is not the subject of
analysis.<br> Metadata must contain the following information:</p>
<pre><code>- The first column is the sample name used in the raw count data.（ex. accession number）
- The second column is the corresponding sample name that matches the sample name in the first column. (e.g. Control_1)
- The third and subsequent columns do not affect the analysis.</code></pre>
<p><img src="www/input_format2.png" class="class" style="width:50.0%"
alt="unchanged image" /> <br><br></p>
<hr />
</div>
<div id="species-1" class="section level4">
<h4>Species</h4>
<p>The following analysis is performed by selecting the dataset
species.</p>
<pre><code>- Conversion to gene symbol if the gene name is ENSEMBL ID
- Performing enrichment analysis</code></pre>
<p><br></p>
</div>
<div id="cut-off-conditions-1" class="section level4">
<h4>Cut-off conditions</h4>
<p>Three types of thresholds, <strong>Fold change</strong>,
<strong>FDR</strong>, and <strong>base mean</strong>, can be set. <br>
<br></p>
</div>
<div id="option-normalized-count-input-1" class="section level4">
<h4>Option: normalized count input</h4>
<p>Base mean cut-off can be set with the uploaded normalized count data
such as TPM counts.<br> The y-axis of the boxplot can be displayed with
the uploaded normalized count data.<br> Note: Uploading raw count data
is not unnecessary. <br> <br></p>
<hr />
</div>
</div>
</div>
<div id="output-1" class="section level2">
<h2>Output</h2>
<div id="input-data-1" class="section level3">
<h3>Input Data</h3>
<p>The uploaded raw count data is displayed.<br> In the case of the “Raw
count data + metadata” format, it also displays the raw count data that
is redefined using the uploaded metadata. <br><br> <img
src="manual_figure/cond%203%20input.png" class="class"
style="width:60.0%" alt="cond 3 input" /></p>
<hr />
</div>
<div id="result-overview-1" class="section level3">
<h3>Result overview</h3>
<p>Three types of clustering analysis are performed: Principal Component
Analysis (PCA), Multi-Dimensional Scaling (MDS), and hierarchical
clustering with the ward.D2.<br> EBSeq multi-comparison analysis is
performed to detect DEGs. Scatter plot and heatmap are displayed as a
result of the DEG analysis.<br> By visualizing DEGs in all three
conditions, the characteristics of gene expression in each condition can
be extracted.<br> The table data of the DEG analysis is
displayed.<br><br></p>
<p><img src="manual_figure/cond%203%20overview1.png" class="class"
style="width:80.0%" alt="cond 3 overview" /><br> <img
src="manual_figure/cond%203%20overview2.png" class="class"
style="width:80.0%" alt="cond 3 overview" /><br> <img
src="manual_figure/cond%203%20overview3.png" class="class"
style="width:80.0%" alt="cond 3 overview" /> <br></p>
<hr />
</div>
<div id="goi-genes-of-interest-profiling-1" class="section level3">
<h3>GOI (Genes of interest) profiling</h3>
<p>The heatmap and boxplot of the genes selected from the GOI list are
displayed.<br><br> <img src="manual_figure/cond%203%20goi1.png"
class="class" style="width:80.0%" alt="cond 3 goi" /><br> <img
src="manual_figure/cond%203%20goi2.png" class="class"
style="width:80.0%" alt="cond 3 goi" /><br></p>
<hr />
</div>
<div id="enrichment-analysis-1" class="section level3">
<h3>Enrichment analysis</h3>
<p>Over-representation analysis is performed to identify the function of
DEGs.<br> The gene set can be selected from the following: <br><a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=H">MSigDB
hallmark gene</a> <br> <a href="https://www.kegg.jp">KEGG</a> <br> <a
href="https://reactome.org">Reactome</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/resource/Pathway+Interaction+Database">PID
(Pathway Interaction Database)</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/dataset/Biocarta+Pathways">BioCarta</a>
<br> <a
href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a>
<br> <a href="http://geneontology.org">GO (biological process, cellular
component, and molecular function)</a> <br> <a
href="https://hpo.jax.org/app/">Human phenotype ontology</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(activator)</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(repressor)</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=TFT">Transcription
factor targets</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=MIR">miRNA
target</a> <br></p>
<p>Dotplot and gene-concept network, cnet, plot are displayed as the
result. <br> Enrichment analysis depends on <a
href="https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html">ClusterProfiler</a>.
<br> <img src="manual_figure/cond%203%20enrichment1.png" class="class"
style="width:80.0%" alt="cond 3 enrichment" /> <br></p>
<hr />
</div>
</div>
</div>
<div id="multi-deg" class="section level1">
<h1>Multi DEG</h1>
<p>Multi DEG detects and visualizes differentially expressed genes by
LRT(Likelihood Ratio Test) DESeq2.<br></p>
<div id="setting-2" class="section level2">
<h2>Setting</h2>
<p><img src="manual_figure/multi%20set.png" class="class"
style="width:50.0%" alt="multi set" /><br></p>
<hr />
<div id="input-format-2" class="section level3">
<h3>Input format</h3>
<p>Two types of raw count data formats can be used as input.<br></p>
<div id="raw-count-data-2" class="section level4">
<h4>1. <strong>raw count data</strong> <br></h4>
<p>Analysis can only be performed with raw count data if the following
conditions are fulfilled:</p>
<pre><code>- The replication number is represented by the underline “_”.
- Do not use the underline &quot;_&quot; for anything else. </code></pre>
<p><img src="www/input_format1.png" class="class" style="width:50.0%"
alt="unchanged image" /> <br>　</p>
<hr />
</div>
<div id="raw-count-data-metadata-2" class="section level4">
<h4>2. <strong>raw count data + metadata</strong> <br></h4>
<p>In this formats, you can visualize the result of two-factor multiple
comparison. Metadata must contain the following information:</p>
<pre><code>- The first column is the sample name used in the raw count data.
- In the first column, the replication number is represented by the underline “_”.
- The second column is the corresponding first condition name that matches the sample name in the first column. 
- The third is the corresponding second condition name that matches the sample name in the first column. </code></pre>
<p><img src="www/input_format_multi.png" class="class"
style="width:50.0%" alt="unchanged image" /> <br><br></p>
<hr />
</div>
<div id="fdr-false-discovery-rate-method-1" class="section level4">
<h4>FDR, False Discovery Rate, method</h4>
<p>The following three types of FDR methods can be selected if the DEG
analysis method is DESeq2 or edgeR.</p>
<ul>
<li>BH, Benjamini-Hochberg method (default)</li>
<li>Qvalue, Storey method</li>
<li><a href="https://www.nature.com/articles/nmeth.3885">IHW,
Independent Hypothesis Weighting</a> <br></li>
</ul>
</div>
<div id="species-2" class="section level4">
<h4>Species</h4>
<p>The following analysis is performed by selecting the dataset
species.</p>
<pre><code>- Conversion to gene symbol if the gene name is ENSEMBL ID
- Performing enrichment analysis</code></pre>
<p><br></p>
</div>
<div id="cut-off-conditions-2" class="section level4">
<h4>Cut-off conditions</h4>
<p>Three types of thresholds, <strong>Fold change</strong>,
<strong>FDR</strong>, and <strong>base mean</strong>, can be set. <br>
<br></p>
</div>
<div id="option-normalized-count-input-2" class="section level4">
<h4>Option: normalized count input</h4>
<p>Base mean cut-off can be set with the uploaded normalized count data
such as TPM counts.<br> The y-axis of the boxplot can be displayed with
the uploaded normalized count data.<br> Note: Uploading raw count data
is not unnecessary. <br> <br></p>
<hr />
</div>
</div>
</div>
<div id="output-2" class="section level2">
<h2>Output</h2>
<div id="input-data-2" class="section level3">
<h3>Input Data</h3>
<p>The uploaded raw count data is displayed.<br> In the case of the “Raw
count data + metadata” format, it also displays the raw count data that
is redefined using the uploaded metadata. <br><br> <img
src="manual_figure/multi%20input1.png" class="class" style="width:80.0%"
alt="multi input" /><br> <img src="manual_figure/multi%20input2.png"
class="class" style="width:80.0%" alt="multi input" /><br></p>
<hr />
</div>
<div id="result-overview-2" class="section level3">
<h3>Result overview</h3>
<p>Four types of clustering analysis are performed: Principal Component
Analysis (PCA), Multi-Dimensional Scaling (MDS), hierarchical clustering
with the ward.D2, and UMAP (Uniform Manifold Approximation and
Projection).<br> Heatmap of DEGs is displayed as a result of the DEG
analysis.<br> The table data of the DEG analysis is
displayed.<br><br></p>
<p>n_neighbors in UMAP is the parameter that controls how UMAP balances
local versus global structure in the data.<br> Low values of n_neighbors
will force UMAP to concentrate on very local structure, while large
values will push UMAP to look at larger neighborhoods of each point.<br>
In the case of a small number of samples, n_neighbors should be
small.<br><br> <img src="manual_figure/multi%20overview1.png"
class="class" style="width:80.0%" alt="multi overview" /><br> <img
src="manual_figure/multi%20overview2.png" class="class"
style="width:80.0%" alt="multi overview" /><br></p>
<hr />
</div>
<div id="divisive-clustering" class="section level3">
<h3>Divisive clustering</h3>
<p>Hierarchical clustering by a top-down clustering approach is
performed.<br> The cluster number is automatically determined.<br>
Boxplot showing gene expression pattern of each cluster is
displayed.<br> The table data showing the relationship between gene and
clusters is displayed.<br> Over-representation analysis is performed to
identify the function of each cluster.<br> The gene set can be selected
from the following: <br><a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=H">MSigDB
hallmark gene</a> <br> <a href="https://www.kegg.jp">KEGG</a> <br> <a
href="https://reactome.org">Reactome</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/resource/Pathway+Interaction+Database">PID
(Pathway Interaction Database)</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/dataset/Biocarta+Pathways">BioCarta</a>
<br> <a
href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a>
<br> <a href="http://geneontology.org">GO (biological process, cellular
component, and molecular function)</a> <br> <a
href="https://hpo.jax.org/app/">Human phenotype ontology</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(activator)</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(repressor)</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=TFT">Transcription
factor targets</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=MIR">miRNA
target</a> <br></p>
<p>Dotplot and gene-concept network, cnet, plot are displayed as the
result. <br> The group displayed by the cnet plot can be switched by
selecting from the “Group” list.<br> Enrichment analysis depends on <a
href="https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html">ClusterProfiler</a>.
<br><br></p>
<p><strong>Select a pair for fold change cut-off</strong><br> The Fold
Change between the selected two groups is subject to “the Cut-off
condition” in the setting panel.<br> This function is effective when you
want to extract genes that change remarkably only between specific
groups.<br></p>
<p><strong>Most significant genes</strong><br> The DEG after cut-off is
extracted in ascending order of FDR, and clustering analysis is
performed.<br> If there are too many DEGs after cut-off, there is a risk
of freezing due to server memory, so narrow down the genes to be
analyzed.<br></p>
<p><img src="manual_figure/multi%20divisive.png" class="class"
style="width:80.0%" alt="multi deg" /><br> <img
src="manual_figure/multi%20divisive1.png" class="class"
style="width:80.0%" alt="multi deg" /><br> <img
src="manual_figure/multi%20divisive2.png" class="class"
style="width:80.0%" alt="multi deg" /><br> <img
src="manual_figure/multi%20divisive3.png" class="class"
style="width:80.0%" alt="multi deg" /><br> <img
src="manual_figure/multi%20divisive4.png" class="class"
style="width:80.0%" alt="multi deg" /><br> <img
src="manual_figure/multi%20divisive5.png" class="class"
style="width:80.0%" alt="multi deg" /><br></p>
<hr />
</div>
<div id="k-means-clustering" class="section level3">
<h3>k-means clustering</h3>
<p>k-means clustering is performed.<br> The number of clusters can be
changed by operating the slide bar.<br> Boxplot showing gene expression
pattern of each cluster is displayed.<br> The table data showing the
relationship between gene and clusters is displayed.<br></p>
<p>Over-representation analysis is performed to identify the function of
each cluster.<br></p>
<p><strong>Select a pair for fold change cut-off</strong><br> The Fold
Change between the selected two groups is subject to “the Cut-off
condition” in the setting panel.<br> This function is effective when you
want to extract genes that change remarkably only between specific
groups.<br></p>
<p><strong>Most significant genes</strong><br> The DEG after cut-off is
extracted in ascending order of FDR, and clustering analysis is
performed.<br> If there are too many DEGs after cut-off, there is a risk
of freezing due to server memory, so narrow down the genes to be
analyzed.<br></p>
<p><img src="manual_figure/multi%20kmeans1.png" class="class"
style="width:80.0%" alt="multi kmeans" /><br> <img
src="manual_figure/multi%20kmeans2.png" class="class"
style="width:80.0%" alt="multi kmeans" /><br> <img
src="manual_figure/multi%20kmeans3.png" class="class"
style="width:80.0%" alt="multi kmeans" /><br> <img
src="manual_figure/multi%20kmeans4.png" class="class"
style="width:80.0%" alt="multi kmeans" /><br> <img
src="manual_figure/multi%20kmeans5.png" class="class"
style="width:80.0%" alt="multi kmeans" /><br> <img
src="manual_figure/multi%20kmeans6.png" class="class"
style="width:80.0%" alt="multi kmeans" /><br></p>
<hr />
</div>
<div id="gsea" class="section level3">
<h3>GSEA</h3>
<p>GSEA (gene set enrichment analysis) is performed to identify the
function of each cluster.<br> The gene set can be selected from the
following: <br><a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=H">MSigDB
hallmark gene</a> <br> <a href="https://www.kegg.jp">KEGG</a> <br> <a
href="https://reactome.org">Reactome</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/resource/Pathway+Interaction+Database">PID
(Pathway Interaction Database)</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/dataset/Biocarta+Pathways">BioCarta</a>
<br> <a
href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a>
<br> <a href="http://geneontology.org">GO (biological process, cellular
component, and molecular function)</a> <br> <a
href="https://hpo.jax.org/app/">Human phenotype ontology</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(activator)</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(repressor)</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=TFT">Transcription
factor targets</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=MIR">miRNA
target</a> <br></p>
<p>Enrichment analysis depends on <a
href="https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html">ClusterProfiler</a>.
<br><br></p>
<p><img src="manual_figure/multi%20GSEA.png" class="class"
style="width:80.0%" alt="multi GSEA" /> <br></p>
<hr />
</div>
</div>
</div>
<div id="venn-diagram" class="section level1">
<h1>Venn diagram</h1>
<p>Venn diagram detects and visualizes the overlap between DEGs from
multiple datasets.<br> By combining with the count data, you can extract
the count data of the common part and create a heatmap.<br><br></p>
<div id="setting-3" class="section level2">
<h2>Setting</h2>
<p><img src="manual_figure/venn%20setting.png" class="class"
style="width:50.0%" alt="venn setting" /><br><br></p>
<hr />
<div id="input-format-3" class="section level3">
<h3>Input format</h3>
<div id="gene-list" class="section level4">
<h4>Gene list</h4>
<p>The format of the gene list used as input is as follows.</p>
<pre><code>- The first column is gene name.
- The second and subsequent columns do not affect the analysis.
- The maximum number of uploads is 7 files.
- File names do not use &quot;.&quot; other than the extension.</code></pre>
<p>There is no problem even if count data is included because the second
and subsequent columns do not affect the analysis.<br> Venn diagram
analysis from <strong>Up to seven files</strong> is performed, and the
gene list including each intersection of the Venn diagram can be
obtained.<br><br> <img src="www/venn_input.png" class="class"
style="width:15.0%" alt="venn input" /><br><br></p>
</div>
<div id="multiple-count-filesfor-integrated-heatmap"
class="section level4">
<h4>Multiple count files（for integrated heatmap）</h4>
<p>Multiple count data is integrated, and then the expression of the
gene set identified by the Venn diagram is visualized.<br> Count data
must fulfill the following condition.</p>
<pre><code>- Use a normalized count file, not raw count file.
- File names do not use &quot;.&quot; other than the extension.</code></pre>
<p><br> <img src="www/input_format1.png" class="class"
style="width:50.0%" alt="venn input" /><br><br></p>
<hr />
</div>
<div id="option-pre-zscoring" class="section level4">
<h4>Option: pre-zscoring</h4>
<p>If True, each count data is z-scored before integrating multiple
count data.<br> <br> <img src="www/pre-zscoring.png" class="class"
style="width:50.0%" alt="venn input" /><br><br></p>
<hr />
</div>
</div>
</div>
<div id="output-3" class="section level2">
<h2>Output</h2>
<div id="input-gene-list" class="section level3">
<h3>Input gene list</h3>
<p>The Venn diagram created from the uploaded gene list is
displayed.<br> For all genes included in the uploaded list, the group to
which the Venn diagram belongs is displayed as a list.<br> <strong>How
to read the gene list</strong><br> The group displayed in the Venn
diagram is created based on the file name of the uploaded gene list (in
the case of the below, day0.txt, day1.txt, and day5.txt are
uploaded).<br> The common part is represented by a colon “:” (in the
case of below, <strong>day0:day1:day5</strong> = “common part of day0,
day1, day5”, <strong>day0:day1</strong> = “common part of day0 and day1,
but not day5”, and <strong>day0</strong> = “unique part of day0”.） <br>
<img src="manual_figure/venn%20venn%20diagram1.png" class="class"
style="width:80.0%" alt="venn venn diagram" /><br> <img
src="manual_figure/venn%20venn%20diagram2.png" class="class"
style="width:80.0%" alt="venn venn diagram" /><br></p>
<hr />
</div>
<div id="integrated-heatmap" class="section level3">
<h3>Integrated heatmap</h3>
<p>Multiple count data is integrated and then genes of the intersection
of the Venn diagram are visualized by heatmap.<br> The intersection to
be visualized can be switched by selecting from “gene list”.<br> Two
types of integrated count tables are output.<br> integrated normalized
count is the result of combining multiple count data without the
z-scoring.<br> Integrated zscored normalized count is z-scored count
data. If pre-zscoring is TRUE, it will be z-scored before combining
multiple data, and if pre-zscoring is False, it will be z-scored after
combining the data.<br> By selecting the ‘not selected’ from the “gene
list”, it is also possible to obtain the integrated count data that
contains all genes. <br> <img
src="manual_figure/venn%20integrated%20heatmap.png" class="class"
style="width:80.0%" alt="venn integrated heatmap" /><br> <img
src="manual_figure/venn%20integrated%20heatmap1.png" class="class"
style="width:80.0%" alt="venn integrated heatmap" /><br> <img
src="manual_figure/venn%20integrated%20heatmap2.png" class="class"
style="width:80.0%" alt="venn integrated heatmap" /><br></p>
<hr />
</div>
<div id="enrichment-analysis-2" class="section level3">
<h3>Enrichment analysis</h3>
<p>Over-representation analysis is performed to identify the function of
each cluster.<br> The gene set can be selected from the following:
<br><a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=H">MSigDB
hallmark gene</a> <br> <a href="https://www.kegg.jp">KEGG</a> <br> <a
href="https://reactome.org">Reactome</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/resource/Pathway+Interaction+Database">PID
(Pathway Interaction Database)</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/dataset/Biocarta+Pathways">BioCarta</a>
<br> <a
href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a>
<br> <a href="http://geneontology.org">GO (biological process, cellular
component, and molecular function)</a> <br> <a
href="https://hpo.jax.org/app/">Human phenotype ontology</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(activator)</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(repressor)</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=TFT">Transcription
factor targets</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=MIR">miRNA
target</a> <br></p>
<p>Dotplot and gene-concept network, cnet, plot are displayed as the
result. <br> The group displayed by the cnet plot can be switched by
selecting from the “Group” list.<br> Enrichment analysis depends on <a
href="https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html">ClusterProfiler</a>.
<br><br> <img src="manual_figure/venn%20enrichment1.png" class="class"
style="width:80.0%" alt="venn enrichment" /> <br> <img
src="manual_figure/venn%20enrichment.png" class="class"
style="width:80.0%" alt="venn enrichment" /> <br></p>
<hr />
</div>
</div>
</div>
<div id="normalized-count-analysis" class="section level1">
<h1>Normalized count analysis</h1>
<p>Normalized count analysis identifies similar samples and gene
expression patterns by clustering methods.<br> By uploading the gene
list, you can extract only the count data of the gene of interest, and
then perform clustering analysis.<br> k-means clustering identifies gene
groups that show similar expression patterns. <br><br></p>
<div id="setting-4" class="section level2">
<h2>Setting</h2>
<p><img src="manual_figure/norm%20set.png" class="class"
style="width:50.0%" alt="norm set" /><br></p>
<hr />
<div id="input-format-4" class="section level3">
<h3>Input format</h3>
<p>The input format for normalized count analysis is the same as for
pair-wise DEG and 3 conditions DEG.<br> If you have read the description
of pair-wise DEG or 3 conditions DEG, you do not need to read the
below.<br><br></p>
<div id="raw-count-data-3" class="section level4">
<h4>1. <strong>raw count data</strong> <br></h4>
<p>Analysis can only be performed with raw count data if the following
conditions are fulfilled:</p>
<pre><code>- The replication number is represented by the underline “_”.
- Do not use the underline &quot;_&quot; for anything else. </code></pre>
<p><img src="www/input_format1.png" class="class" style="width:0.0%"
alt="unchanged image" /> <br>　</p>
<hr />
</div>
<div id="raw-count-data-metadata-3" class="section level4">
<h4>2. <strong>raw count data + metadata</strong> <br></h4>
<p>You can use this format if the above conditions are not fulfilled,
for example, if the sample name is an accession number or if the raw
count data contains extra information that is not the subject of
analysis.<br> Metadata must contain the following information:</p>
<pre><code>- The first column is the sample name used in the raw count data.（ex. accession number）
- The second column is the corresponding sample name that matches the sample name in the first column. (e.g. Control_1)
- The third and subsequent columns do not affect the analysis.</code></pre>
<p><img src="www/input_format2.png" class="class" style="width:50.0%"
alt="unchanged image" /> <br><br></p>
<hr />
</div>
<div id="species-3" class="section level4">
<h4>Species</h4>
<p>The following analysis is performed by selecting the dataset
species.</p>
<pre><code>- Conversion to gene symbol if the gene name is ENSEMBL ID</code></pre>
<p><br><br></p>
</div>
<div id="filter-option-1" class="section level4">
<h4>Filter option 1</h4>
<p>In the case that the gene list is uploaded, clustering analysis of
the normalized count data is performed after the gene extraction with
the gene list. <br> <br></p>
</div>
<div id="filter-option-2" class="section level4">
<h4>Filter option 2</h4>
<p>The base mean threshold can be set. <br> <br></p>
<hr />
</div>
</div>
</div>
<div id="output-4" class="section level2">
<h2>Output</h2>
<div id="input-normalized-count-data" class="section level3">
<h3>Input Normalized Count Data</h3>
<p>The uploaded files are displayed.<br> In the case that metadata and
gene list are uploaded, it also show the normalized count data that is
re-defined with metadata and gene list. <br> <img
src="manual_figure/norm%20input.png" class="class" style="width:80.0%"
alt="norm input" /> <br></p>
<hr />
</div>
<div id="clustering" class="section level3">
<h3>Clustering</h3>
<p>Three types of clustering analysis are performed: Principal Component
Analysis (PCA), Multi-Dimensional Scaling (MDS), and hierarchical
clustering with the ward.D2.<br></p>
<p><img src="manual_figure/norm%20overview1.png" class="class"
style="width:80.0%" alt="norm overview" /><br> <img
src="manual_figure/norm%20overview2.png" class="class"
style="width:80.0%" alt="norm overview" /><br></p>
<hr />
</div>
<div id="goi-genes-of-interest-profiling-2" class="section level3">
<h3>GOI (Genes of interest) profiling</h3>
<p>The heatmap and boxplot of the genes selected from the GOI list are
displayed.<br><br></p>
<p><br> <img src="manual_figure/norm%20goi1.png" class="class"
style="width:80.0%" alt="norm goi" /><br> <img
src="manual_figure/norm%20goi2.png" class="class" style="width:80.0%"
alt="norm goi" /><br></p>
<hr />
</div>
<div id="k-means-clustering-1" class="section level3">
<h3>k-means clustering</h3>
<p>k-means clustering is performed.<br> The number of clusters can be
changed by operating the slide bar.<br> The table data showing the
relationship between gene and clusters is displayed.</p>
<p><br> <img src="manual_figure/norm%20kmeans1.png" class="class"
style="width:80.0%" alt="norm kmeans" /><br> <img
src="manual_figure/norm%20kmeans2.png" class="class" style="width:80.0%"
alt="norm kmeans" /><br> <img src="manual_figure/norm%20kmeans3.png"
class="class" style="width:80.0%" alt="norm kmeans" /><br></p>
<hr />
</div>
</div>
</div>
<div id="enrichment-viewer" class="section level1">
<h1>Enrichment viewer</h1>
<p>Enrichment viewer determines and visualizes biological functions of
gene set of interest.<br> By using the results of the Venn diagram and
k-means clustering, the functions of the identified gene groups can be
compared and visualized. <br><br></p>
<div id="setting-5" class="section level2">
<h2>Setting</h2>
<p><img src="manual_figure/enrich%20set.png" class="class"
style="width:50.0%" alt="enrich set" /><br><br><br><br></p>
<hr />
<div id="input-format-5" class="section level3">
<h3>Input format</h3>
<p>The format of the gene list used as input is as follows.<br></p>
<pre><code>- The first column is gene name
- The second column is the corresponding group name that matches the gene name in the first column.
- The third and subsequent columns do not affect the analysis.</code></pre>
<p>There is no problem even if count data is included because the third
and subsequent columns do not affect the analysis.<br> You can use the
table data obtained by the Venn diagram or k-means clustering without
changing the format.<br><br></p>
<hr />
<div id="species-4" class="section level4">
<h4>Species</h4>
<p>The following analysis is performed by selecting the dataset
species.</p>
<pre><code>- Conversion to gene symbol if the gene name is ENSEMBL ID
- Performing enrichment analysis</code></pre>
<p><br></p>
<hr />
</div>
<div id="most-significant-pathways" class="section level4">
<h4>Most significant pathways</h4>
<p>You can change the number of pathways displayed in the graphs.
<br></p>
<hr />
</div>
</div>
</div>
<div id="output-5" class="section level2">
<h2>Output</h2>
<div id="input-gene-list-for-enrichment-analysis"
class="section level3">
<h3>Input gene list for enrichment analysis</h3>
<p>The uploaded gene list is displayed.<br></p>
<p><img src="manual_figure/enrich%20input1.png" class="class"
style="width:80.0%" alt="enrich input" /> <br></p>
<hr />
</div>
<div id="enrichment-analysis-3" class="section level3">
<h3>Enrichment analysis</h3>
<p>Over-representation analysis is performed to identify the function of
gene sets.<br> The gene set can be selected from the following:<br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=H">MSigDB
hallmark gene</a> <br> <a href="https://www.kegg.jp">KEGG</a> <br> <a
href="https://reactome.org">Reactome</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/resource/Pathway+Interaction+Database">PID
(Pathway Interaction Database)</a> <br> <a
href="https://maayanlab.cloud/Harmonizome/dataset/Biocarta+Pathways">BioCarta</a>
<br> <a
href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a>
<br> <a href="http://geneontology.org">GO (biological process, cellular
component, and molecular function)</a> <br> <a
href="https://hpo.jax.org/app/">Human phenotype ontology</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(activator)</a> <br> <a
href="https://saezlab.github.io/dorothea/">DoRothEA regulon
(repressor)</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=TFT">Transcription
factor targets</a> <br> <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=MIR">miRNA
target</a> <br></p>
<p>Dotplot and gene-concept network, cnet, plot are displayed as the
result. <br> The group displayed by the cnet plot can be switched by
selecting from the “Group” list.<br> Enrichment analysis depends on <a
href="https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html">ClusterProfiler</a>.
<br> <img src="manual_figure/enrich%20enrichment1.png" class="class"
style="width:80.0%" alt="enrich enrichment" /><br> <img
src="manual_figure/enrich%20enrichment2.png" class="class"
style="width:80.0%" alt="enrich enrichment" /><br></p>
<hr />
</div>
<div id="promoter-motif-analysis" class="section level3">
<h3>Promoter motif analysis</h3>
<p>Promoter motif analysis identifies DNA-binding motifs that enrich the
promoter region of each gene group.<br> Species: human (hg19) or mouse
(mm10)<br> Upstream: upstream range from the TSS <br> Downstream:
downstream range from the TSS<br> padj: threshold for dotplot<br> Press
“Start” button to start the analysis.<br><br></p>
<p>By selecting a motif of interest from the “Motif table”, you can
display the region where Motif exists (limited to the specified TSS
region of the enriched gene group) in the “Motif region”.<br> This
analysis depends on <a
href="https://fmicompbio.github.io/monaLisa/articles/monaLisa.html">monaLisa</a>.<br></p>
<p><img src="manual_figure/enrich%20promoter1.png" class="class"
style="width:80.0%" alt="enrich promoter" /><br> <img
src="manual_figure/enrich%20promoter2.png" class="class"
style="width:80.0%" alt="enrich promoter" /><br> <img
src="manual_figure/enrich%20promoter3.png" class="class"
style="width:80.0%" alt="enrich promoter" /><br></p>
<hr />
</div>
</div>
</div>
<div id="more" class="section level1">
<h1>More</h1>
<div id="volcano-navi" class="section level2">
<h2>Volcano navi</h2>
<p>Volcano navi shows volcano plot.<br> The input format is as
follows.</p>
<pre><code>- The first column is gene name.
- A file includes columns named &quot;log2FoldChange&quot; and &quot;padj&quot;.
- Other columns do not affect the analysis.</code></pre>
<p>You can use not only RNA-seq data but also the other omics data.<br>
By selecting genes from the GOI list, you can label only the GOI among
the points displayed in the volcano plot.<br> The x-axis and y-axis
ranges can be freely changed by operating the slide bar.<br> In the case
that the normalized count data is uploaded, the heatmap and boxplot of
the genes selected from the GOI list are displayed.<br></p>
<p><img src="manual_figure/volcano%20navi.png" class="class"
style="width:100.0%" alt="volcano navi" /><br></p>
<hr />
</div>
<div id="msigdb-gene-set" class="section level2">
<h2>MSigDB gene set</h2>
<p>The gene sets registered with <a
href="http://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp">MSigDB
(Molecular Signatures Data Base)</a> can be downloaded in this
section.<br></p>
<p><img src="manual_figure/MSigDB.png" class="class"
style="width:100.0%" alt="MSigDB" /><br></p>
<hr />
</div>
<div id="dorothea-regulon" class="section level2">
<h2>DoRothEA regulon</h2>
<p><a href="https://saezlab.github.io/dorothea/">DoRothEA (Discriminant
Regulon Expression Analysis)</a> allows you to browse the data of
DoRothEA.<br> By entering the gene name in “Target name for TF search”
and “TF name for target search”, bi-directional search of “transcription
factor ⇄ target gene” is possible.<br> <strong>Confidence</strong> <br>
The official page recommends data with reliable <strong>confidence
A-C</strong>.<br> Enrichment analysis by RNAseqChef uses data with
<strong>confidence A-C</strong> to increase the accuracy of
transcription factor prediction.</p>
<p><img src="manual_figure/DoRothEA1.png" class="class"
style="width:100.0%" alt="DoRothEA" /><br> <img
src="manual_figure/DoRothEA2.png" class="class" style="width:100.0%"
alt="DoRothEA" /><br></p>
<hr />
</div>
<div id="reference" class="section level2">
<h2>Reference</h2>
<p>Shiny framework</p>
<ul>
<li>Winston Chang, Joe Cheng, JJ Allaire, Carson Sievert, Barret
Schloerke, Yihui Xie, Jeff Allen, Jonathan McPherson, Alan Dipert and
Barbara Borges (2021). shiny: Web Application Framework for R. R package
version 1.7.1. <a href="https://CRAN.R-project.org/package=shiny"
class="uri">https://CRAN.R-project.org/package=shiny</a></li>
<li>Eric Bailey (2022). shinyBS: Twitter Bootstrap Components for Shiny.
R package version 0.61.1. <a
href="https://CRAN.R-project.org/package=shinyBS"
class="uri">https://CRAN.R-project.org/package=shinyBS</a></li>
<li>Yihui Xie, Joe Cheng and Xianying Tan (2022). DT: A Wrapper of the
JavaScript Library ‘DataTables’. R package version 0.23. <a
href="https://CRAN.R-project.org/package=DT"
class="uri">https://CRAN.R-project.org/package=DT</a></li>
</ul>
<p>EBSeq (for ebseq)</p>
<ul>
<li>Ning Leng and Christina Kendziorski (2020). EBSeq: An R package for
gene and isoform differential expression analysis of RNA-seq data. R
package version 1.30.0.</li>
</ul>
<p>DESeq2 (for deseq2)</p>
<ul>
<li>Love, M.I., Huber, W., Anders, S. Moderated estimation of fold
change and dispersion for RNA-seq data with DESeq2 Genome Biology
15(12):550 (2014)</li>
</ul>
<p>edgeR (for edger)</p>
<ul>
<li>Robinson MD, McCarthy DJ and Smyth GK (2010). edgeR: a Bioconductor
package for differential expression analysis of digital gene expression
data. Bioinformatics 26, 139-140</li>
</ul>
<p>IHW, Independent hypothesis weighting, and qvalue (for fdr control
method of deseq2 and edger)</p>
<ul>
<li>Nikolaos Ignatiadis, Bernd Klaus, Judith Zaugg and Wolfgang Huber
(2016): Data-driven hypothesis weighting increases detection power in
genome-scale multiple testing. Nature Methods 13:577, doi:
10.1038/nmeth.3885</li>
<li>John D. Storey, Andrew J. Bass, Alan Dabney and David Robinson
(2021). qvalue: Q-value estimation for false discovery rate control. R
package version 2.26.0. <a href="http://github.com/jdstorey/qvalue"
class="uri">http://github.com/jdstorey/qvalue</a></li>
</ul>
<p>ggdendro (for dendrograms)</p>
<ul>
<li>Andrie de Vries and Brian D. Ripley (2020). ggdendro: Create
Dendrograms and Tree Diagrams Using ‘ggplot2’. R package version 0.1.22.
<a href="https://CRAN.R-project.org/package=ggdendro"
class="uri">https://CRAN.R-project.org/package=ggdendro</a></li>
</ul>
<p>umap (for UMAP)</p>
<ul>
<li>Konopka T (2022). <em>umap: Uniform Manifold Approximation and
Projection</em>. R package version 0.2.8.0, <a
href="https://CRAN.R-project.org/package=umap"
class="uri">https://CRAN.R-project.org/package=umap</a>.</li>
</ul>
<p>clusterProfiler, DOSE, msigdbr, dorothea (for enrichment
analysis)</p>
<ul>
<li>T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L
Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal
enrichment tool for interpreting omics data. The Innovation. 2021,
2(3):100141</li>
<li>Guangchuang Yu, Li-Gen Wang, Guang-Rong Yan, Qing-Yu He. DOSE: an
R/Bioconductor package for Disease Ontology Semantic and Enrichment
analysis. Bioinformatics 2015 31(4):608-609</li>
<li>Dolgalev I (2022). <em>msigdbr: MSigDB Gene Sets for Multiple
Organisms in a Tidy Data Format</em>. R package version 7.5.1, <a
href="https://CRAN.R-project.org/package=msigdbr"
class="uri">https://CRAN.R-project.org/package=msigdbr</a>.</li>
<li>Garcia-Alonso L, Holland CH, Ibrahim MM, Turei D, Saez-Rodriguez J.
‘Benchmark and integration of resources for the estimation of human
transcription factor activities.’ Genome Research. 2019. DOI:
10.1101/gr.240663.118.</li>
</ul>
<p>AnnotationDbi, org.Hs.eg.db, org.Mm.eg.db, org.Rn.eg.db,
org.Xl.eg.db, org.Dm.eg.db, and org.Ce.eg.db (for genome wide
annotation)</p>
<ul>
<li>Hervé Pagès, Marc Carlson, Seth Falcon and Nianhua Li (2020).
AnnotationDbi: Manipulation of SQLite-based annotations in Bioconductor.
R package version 1.52.0. <a
href="https://bioconductor.org/packages/AnnotationDbi"
class="uri">https://bioconductor.org/packages/AnnotationDbi</a></li>
<li>Marc Carlson (2020). org.Hs.eg.db: Genome wide annotation for Human.
R package version 3.12.0.</li>
<li>Marc Carlson (2020). org.Mm.eg.db: Genome wide annotation for Mouse.
R package version 3.12.0.</li>
<li>Marc Carlson (2022). org.Rn.eg.db: Genome wide annotation for Rat. R
package version 3.15.0.</li>
<li>Marc Carlson (2022). org.Xl.eg.db: Genome wide annotation for Worm.
R package version 3.15.0.</li>
<li>Marc Carlson (2022). org.Dm.eg.db: Genome wide annotation for Rat. R
package version 3.15.0.</li>
<li>Marc Carlson (2022). org.Ce.eg.db: Genome wide annotation for Worm.
R package version 3.15.0.</li>
</ul>
<p>genefilter (for z-score normalization)</p>
<ul>
<li>R. Gentleman, V. Carey, W. Huber and F. Hahne (2021). genefilter:
methods for filtering genes from high-throughput experiments. R package
version 1.72.1.</li>
</ul>
<p>ComplexHeatmap (for heatmap and k-means clustering)</p>
<ul>
<li>Gu, Z. (2016) Complex heatmaps reveal patterns and correlations in
multidimensional genomic data. Bioinformatics.</li>
</ul>
<p>ggplot2 and ggpubr (for boxplot and scater plot)</p>
<ul>
<li>H. Wickham. ggplot2: Elegant Graphics for Data Analysis.
Springer-Verlag New York, 2016.</li>
<li>Alboukadel Kassambara (2020). ggpubr: ‘ggplot2’ Based Publication
Ready Plots. R package version 0.4.0. <a
href="https://CRAN.R-project.org/package=ggpubr"
class="uri">https://CRAN.R-project.org/package=ggpubr</a></li>
</ul>
<p>DEGreport (for divisive clustering analysis)</p>
<ul>
<li>Pantano L (2022). DEGreport: Report of DEG analysis. R package
version 1.32.0, <a href="http://lpantano.github.io/DEGreport"
class="uri">http://lpantano.github.io/DEGreport</a></li>
</ul>
<p>venn (for venn diagram analysis)</p>
<ul>
<li>Adrian Dusa (2021). venn: Draw Venn Diagrams. R package version
1.10. <a href="https://CRAN.R-project.org/package=venn"
class="uri">https://CRAN.R-project.org/package=venn</a></li>
</ul>
<p>dplyr and tidyr (for data manipulation)</p>
<ul>
<li>Hadley Wickham, Romain François, Lionel Henry and Kirill Müller
(2021). dplyr: A Grammar of Data Manipulation. R package version 1.0.7.
<a href="https://CRAN.R-project.org/package=dplyr"
class="uri">https://CRAN.R-project.org/package=dplyr</a></li>
<li>Hadley Wickham (2021). tidyr: Tidy Messy Data. R package version
1.1.3. <a href="https://CRAN.R-project.org/package=tidyr"
class="uri">https://CRAN.R-project.org/package=tidyr</a></li>
</ul>
<p>sample data</p>
<ul>
<li>R. Takase, S. Hino, K. Nagaoka, K. Anan, K. Kohrogi, H. Araki, Y.
Hino, A. Sakamoto, T. B. Nicholson, T. Chen, M. Nakao, Lysine-specific
demethylase-2 is distinctively involved in brown and beige adipogenic
differentiation. FASEB J. 33, 5300–5311 (2019).</li>
<li>Saunderson EA, Stepper P, Gomm JJ, Hoa L et al. Hit-and-run
epigenetic editing prevents senescence entry in primary breast cells
from healthy donors. Nat Commun 2017 Nov 13;8(1):1450. PMID:
29133799</li>
</ul>
<hr />
</div>
</div>
<div id="license" class="section level1">
<h1>License</h1>
<p>This shiny code is licensed under the GPLv3. Please see the file
LICENSE.md for information.</p>
<pre><code>RNAseqChef, an RNA-seq data controller highlighting gene expression features
Shiny App for automated, systematic, and integrated RNA-seq differential expression analysis
Copyright (C) 2022  Kan Etoh

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.

You may contact the author of this code, Kan Etoh, at &lt;kaneto@kumamoto-u.ac.jp&gt;</code></pre>
<hr />
</div>
<div id="session-info" class="section level1">
<h1>Session info</h1>
<pre class="r"><code> sessionInfo()</code></pre>
<pre><code>## R version 4.2.1 (2022-06-23)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur ... 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] ja_JP.UTF-8/ja_JP.UTF-8/ja_JP.UTF-8/C/ja_JP.UTF-8/ja_JP.UTF-8
## 
## attached base packages:
##  [1] grid      stats4    tools     stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2  
##  [2] TxDb.Mmusculus.UCSC.mm10.knownGene_3.10.0
##  [3] GenomicFeatures_1.48.4                   
##  [4] BSgenome.Hsapiens.UCSC.hg19_1.4.3        
##  [5] BSgenome.Mmusculus.UCSC.mm10_1.4.3       
##  [6] BSgenome_1.64.0                          
##  [7] rtracklayer_1.56.1                       
##  [8] Biostrings_2.64.1                        
##  [9] XVector_0.36.0                           
## [10] TFBSTools_1.34.0                         
## [11] JASPAR2020_0.99.10                       
## [12] BiocParallel_1.30.3                      
## [13] monaLisa_1.2.0                           
## [14] biomaRt_2.52.0                           
## [15] umap_0.2.9.0                             
## [16] dorothea_1.8.0                           
## [17] clusterProfiler.dplyr_0.0.2              
## [18] BiocManager_1.30.18                      
## [19] shinyjs_2.1.0                            
## [20] plotly_4.10.0                            
## [21] shinyBS_0.61.1                           
## [22] ComplexHeatmap_2.12.1                    
## [23] genefilter_1.78.0                        
## [24] msigdbr_7.5.1                            
## [25] DOSE_3.22.1                              
## [26] enrichplot_1.16.2                        
## [27] clusterProfiler_4.4.4                    
## [28] org.Xl.eg.db_3.15.0                      
## [29] org.Ce.eg.db_3.15.0                      
## [30] org.Dm.eg.db_3.15.0                      
## [31] org.Rn.eg.db_3.15.0                      
## [32] org.Mm.eg.db_3.15.0                      
## [33] org.Hs.eg.db_3.15.0                      
## [34] AnnotationDbi_1.58.0                     
## [35] DEGreport_1.32.0                         
## [36] qvalue_2.28.0                            
## [37] IHW_1.24.0                               
## [38] edgeR_3.38.4                             
## [39] limma_3.52.3                             
## [40] ggnewscale_0.4.7                         
## [41] EBSeq_1.36.0                             
## [42] testthat_3.1.4                           
## [43] gplots_3.1.3                             
## [44] blockmodeling_1.1.3                      
## [45] DESeq2_1.36.0                            
## [46] SummarizedExperiment_1.26.1              
## [47] Biobase_2.56.0                           
## [48] MatrixGenerics_1.8.1                     
## [49] matrixStats_0.62.0                       
## [50] GenomicRanges_1.48.0                     
## [51] GenomeInfoDb_1.32.4                      
## [52] IRanges_2.30.1                           
## [53] S4Vectors_0.34.0                         
## [54] BiocGenerics_0.42.0                      
## [55] cowplot_1.1.1                            
## [56] gridExtra_2.3                            
## [57] ggplotify_0.1.0                          
## [58] ggdendro_0.1.23                          
## [59] ggrepel_0.9.1                            
## [60] venn_1.11                                
## [61] ggpubr_0.4.0                             
## [62] forcats_0.5.2                            
## [63] stringr_1.4.1                            
## [64] dplyr_1.0.10                             
## [65] purrr_0.3.4                              
## [66] readr_2.1.2                              
## [67] tidyr_1.2.1                              
## [68] tibble_3.1.8                             
## [69] ggplot2_3.3.6                            
## [70] tidyverse_1.3.2                          
## [71] multcomp_1.4-20                          
## [72] TH.data_1.1-1                            
## [73] MASS_7.3-58.1                            
## [74] survival_3.4-0                           
## [75] mvtnorm_1.1-3                            
## [76] rstatix_0.7.0                            
## [77] gdata_2.18.0.1                           
## [78] DT_0.25                                  
## [79] shiny_1.7.2                              
## 
## loaded via a namespace (and not attached):
##   [1] Rsamtools_2.12.0            foreach_1.5.2              
##   [3] glmnet_4.1-4                crayon_1.5.1               
##   [5] nlme_3.1-159                backports_1.4.1            
##   [7] reprex_2.0.2                GOSemSim_2.22.0            
##   [9] rlang_1.0.6                 readxl_1.4.1               
##  [11] filelock_1.0.2              rjson_0.2.21               
##  [13] bcellViper_1.32.0           CNEr_1.32.0                
##  [15] bit64_4.0.5                 glue_1.6.2                 
##  [17] poweRlaw_0.70.6             parallel_4.2.1             
##  [19] haven_2.5.1                 tidyselect_1.1.2           
##  [21] XML_3.99-0.10               zoo_1.8-11                 
##  [23] GenomicAlignments_1.32.1    xtable_1.8-4               
##  [25] magrittr_2.0.3              evaluate_0.16              
##  [27] sm_2.2-5.7.1                cli_3.4.1                  
##  [29] zlibbioc_1.42.0             rstudioapi_0.14            
##  [31] bslib_0.4.0                 fastmatch_1.1-3            
##  [33] treeio_1.20.2               xfun_0.33                  
##  [35] askpass_1.1                 clue_0.3-61                
##  [37] cluster_2.1.4               caTools_1.18.2             
##  [39] tidygraph_1.2.2             KEGGREST_1.36.3            
##  [41] logging_0.10-108            ape_5.6-2                  
##  [43] TFMPvalue_0.0.8             png_0.1-7                  
##  [45] reshape_0.8.9               withr_2.5.0                
##  [47] bitops_1.0-7                slam_0.1-50                
##  [49] ggforce_0.3.4               plyr_1.8.7                 
##  [51] cellranger_1.1.0            pracma_2.4.2               
##  [53] pillar_1.8.1                GlobalOptions_0.1.2        
##  [55] stabs_0.6-4                 cachem_1.0.6               
##  [57] fs_1.5.2                    GetoptLong_1.0.5           
##  [59] vctrs_0.4.1                 ellipsis_0.3.2             
##  [61] generics_0.1.3              munsell_0.5.0              
##  [63] tweenr_2.0.2                fgsea_1.22.0               
##  [65] DelayedArray_0.22.0         fastmap_1.1.0              
##  [67] compiler_4.2.1              abind_1.4-5                
##  [69] httpuv_1.6.6                GenomeInfoDbData_1.2.8     
##  [71] lattice_0.20-45             utf8_1.2.2                 
##  [73] later_1.3.0                 BiocFileCache_2.4.0        
##  [75] jsonlite_1.8.0              scales_1.2.1               
##  [77] tidytree_0.4.1              carData_3.0-5              
##  [79] lazyeval_0.2.2              promises_1.2.0.1           
##  [81] car_3.1-0                   doParallel_1.0.17          
##  [83] R.utils_2.12.0              reticulate_1.26            
##  [85] rmarkdown_2.16              sandwich_3.0-2             
##  [87] vioplot_0.3.7               downloader_0.4             
##  [89] igraph_1.3.5                yaml_2.3.5                 
##  [91] htmltools_0.5.3             memoise_2.0.1              
##  [93] BiocIO_1.6.0                locfit_1.5-9.6             
##  [95] graphlayouts_0.8.1          viridisLite_0.4.1          
##  [97] digest_0.6.29               assertthat_0.2.1           
##  [99] mime_0.12                   rappdirs_0.3.3             
## [101] RSQLite_2.2.17              yulab.utils_0.0.5          
## [103] data.table_1.14.2           blob_1.2.3                 
## [105] R.oo_1.25.0                 lpsymphony_1.24.0          
## [107] splines_4.2.1               googledrive_2.0.0          
## [109] RCurl_1.98-1.8              broom_1.0.1                
## [111] hms_1.1.2                   modelr_0.1.9               
## [113] colorspace_2.0-3            ConsensusClusterPlus_1.60.0
## [115] mnormt_2.1.0                shape_1.4.6                
## [117] aplot_0.1.7                 sass_0.4.2                 
## [119] Rcpp_1.0.9                  circlize_0.4.15            
## [121] fansi_1.0.3                 tzdb_0.3.0                 
## [123] brio_1.1.3                  Nozzle.R1_1.1-1.1          
## [125] R6_2.5.1                    lifecycle_1.0.2            
## [127] curl_4.3.2                  ggsignif_0.6.3             
## [129] googlesheets4_1.0.1         jquerylib_0.1.4            
## [131] DO.db_2.9                   Matrix_1.5-1               
## [133] RColorBrewer_1.1-3          iterators_1.0.14           
## [135] htmlwidgets_1.5.4           polyclip_1.10-0            
## [137] shadowtext_0.1.2            gridGraphics_0.5-1         
## [139] seqLogo_1.62.0              rvest_1.0.3                
## [141] openssl_2.0.3               patchwork_1.1.2            
## [143] codetools_0.2-18            lubridate_1.8.0            
## [145] GO.db_3.15.0                gtools_3.9.3               
## [147] prettyunits_1.1.1           psych_2.2.5                
## [149] dbplyr_2.2.1                RSpectra_0.16-1            
## [151] R.methodsS3_1.8.2           gtable_0.3.1               
## [153] DBI_1.1.3                   ggfun_0.0.7                
## [155] httr_1.4.4                  KernSmooth_2.23-20         
## [157] stringi_1.7.8               progress_1.2.2             
## [159] reshape2_1.4.4              farver_2.1.1               
## [161] annotate_1.74.0             viridis_0.6.2              
## [163] fdrtool_1.2.17              ggtree_3.4.2               
## [165] xml2_1.3.3                  admisc_0.30                
## [167] restfulr_0.0.15             geneplotter_1.74.0         
## [169] bit_4.0.4                   scatterpie_0.1.8           
## [171] ggraph_2.0.6                pkgconfig_2.0.3            
## [173] babelgene_22.3              gargle_1.2.1               
## [175] DirichletMultinomial_1.38.0 lasso2_1.2-22              
## [177] knitr_1.40</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
